core/src/main/scala/org/apache/spark/util/io/ChunkedByteBuffer.scala

```diff
@@ -84,6 +91,74 @@ private[spark] class ChunkedByteBuffer(var chunks: Array[ByteBuffer]) {
     }
   }
 
+  /**
+   * write to ObjectOutput with zero copy if possible
+   */
+  override def writeExternal(out: ObjectOutput): Unit = {
+    // we want to keep the chunks layout
+    out.writeInt(chunks.length)
+    chunks.foreach(buffer => out.writeInt(buffer.limit()))
+    chunks.foreach(buffer => out.writeBoolean(buffer.isDirect))
+    var buffer: Array[Byte] = null
+    val bufferLen = ChunkedByteBuffer.COPY_BUFFER_LEN
+
+    getChunks().foreach { chunk => {
+      if (chunk.hasArray) {
+        // zero copy if the bytebuffer is backed by bytes array
+        out.write(chunk.array(), chunk.arrayOffset(), chunk.limit())
+      } else {
+        // fallback to copy approach
+        if (buffer == null) {
+          buffer = new Array[Byte](bufferLen)
+        }
+        var bytesToRead = Math.min(chunk.remaining(), bufferLen)
+        while (bytesToRead > 0) {
+          chunk.get(buffer, 0, bytesToRead)
+          out.write(buffer, 0, bytesToRead)
+          bytesToRead = Math.min(chunk.remaining(), bufferLen)
+        }
```

Comment:
nit: Pull this out into a separate method ?
I will need to search our codebase, but I would have expected this snippet to be already there somewhere.

Reply:
There should be... but I was not able to locate them....not very familiar with the code base. Feel free to let me know if you find them

Reply:
There is `Utils.writeByteBuffer` which writes a buffer to an `OutputStream`.
It does not do a very good job, since it allocates a buffer the same size as `remaining` - instead, we should enhance it to do what this method is doing.

Additionally, we can use a `ThreadLocal[Array[Byte]]` in `Utils` for use with this copy (the `buffer` here)

Reply:
I tried to reuse `Utils.writeByteBuffer`, and I noticed that there're two versions of `Utils.writeByteBuffer` for `OutputStream` and `DataOutput` respectively and the code is identical so added a `Utils.writeByteBufferImpl` to extract the common logic, also added ThreadLocal[Array[Byte]]